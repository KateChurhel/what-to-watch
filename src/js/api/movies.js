/* istanbul ignore file */
import { sendGetRequest } from './index';

export const getCategoryData = (category, data) => sendGetRequest(`/${category}/popular`, data);

export const getDetailsData = (category, id, data) => sendGetRequest(`/${category}/${id}`, data);

export const getSearchResult = (category, data) => sendGetRequest(`/search/${category}`, data);

export const getDiscoverResult = (data) => sendGetRequest('/discover/movie', data);

export const getGenres = () => sendGetRequest('/genre/movie/list');
