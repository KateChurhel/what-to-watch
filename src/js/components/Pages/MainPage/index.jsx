// libraries
import React from 'react';
import PropTypes from 'prop-types';
// constants
import CATEGORIES from '../../../constants/categories';
// views
import BigMovieCard from '../../Partials/BigMovieCard';
import Slider from '../../Partials/Slider';

const MainPage = ({ mostPopularMovies = [], mostPopularTvs = [], bigMovieCardData = {} }) => (
  <>
    <BigMovieCard {...bigMovieCardData} />

    <div className="container">
      {mostPopularMovies.length ? (
        <>
          <div className="swiper__title-wrapper">
            <h2 className="swiper__title">Most Popular Movies</h2>
            <a href="/movie/1">Show All</a>
          </div>
          <Slider category={CATEGORIES.movie} items={mostPopularMovies} />
        </>
      ) : null}
      {mostPopularTvs.length ? (
        <>
          <div className="swiper__title-wrapper">
            <h2 className="swiper__title">Most Popular TV Shows</h2>
            <a href="/tv/1">Show All</a>
          </div>
          <Slider category={CATEGORIES.tv} items={mostPopularTvs} />
        </>
      ) : null}
    </div>
  </>
);

MainPage.propTypes = {
  mostPopularMovies: PropTypes.arrayOf(PropTypes.object),
  mostPopularTvs: PropTypes.arrayOf(PropTypes.object),
  bigMovieCardData: PropTypes.shape({}),
};

export default MainPage;
