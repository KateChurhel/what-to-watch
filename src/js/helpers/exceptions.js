// constants
import DEFAULT_ERROR from '../constants/errors';

export class RequestError extends Error {
  constructor(status = '', responseData) {
    super();

    const errorData = responseData || {};

    this.message = errorData.message || DEFAULT_ERROR;
    this.errors = errorData.errors;
    this.status = status;
    this.name = 'RequestError';
  }
}

export default {};
