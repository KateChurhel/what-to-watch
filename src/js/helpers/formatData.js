import moment from 'moment';
import 'moment-duration-format';

export const formatTime = (time, unit, format) => moment.duration(time, unit).format(format, { trim: false });
export const formatDate = (date, format) => moment(date).format(format);

export const parseQueryString = (queryString) => {
  const query = {};
  const pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
  for (let i = 0; i < pairs.length; i++) {
    const pair = pairs[i].split('=');
    query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
  }

  return query;
};

export const getStringValues = (array = [], key) => (array.length ? array.map((item) => item[key]).join(', ') : '');

export const formatCardData = ({
  number_of_episodes: numberOfEpisodes, number_of_seasons: numberOfSeasons, original_name: originalName,
  episode_run_time: episodeRunTime, reviews: { results: reviews = [] } = {}, homepage, original_title: originalTitle,
  original_language: originalLanguage, spoken_languages: spokenLanguages, status, runtime, tagline,
  release_date: releaseDate, genres, poster_path: poster, backdrop_path: backgroundImage,
  vote_average: voteAverage, vote_count: voteCount, overview, title, name, id,
  credits: { cast = [] } = {}, recommendations: { results: recommendations = [] } = {},
} = {}) => {
  const formattedRunTime = formatTime(runtime || episodeRunTime, 'minutes', 'h[h] m[m]');
  const releaseYear = formatDate(releaseDate, 'YYYY');
  const genresList = getStringValues(genres, 'name');
  const spokenLanguagesList = getStringValues(spokenLanguages, 'name');

  return {
    id,
    numberOfEpisodes,
    numberOfSeasons,
    episodeRunTime,
    reviews,
    homepage,
    originalLanguage,
    status,
    tagline,
    releaseDate,
    poster,
    backgroundImage,
    voteAverage,
    voteCount,
    overview,
    cast,
    recommendations,
    originalTitle: originalTitle || originalName,
    cardTitle: title || name,
    formattedRunTime,
    releaseYear,
    genresList,
    spokenLanguagesList,
  };
};
