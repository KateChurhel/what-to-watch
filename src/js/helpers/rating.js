import USER_RATINGS from '../constants/rating';

export const computeRatingLevel = (rating) => USER_RATINGS.filter(
  (rank) => rank.min <= rating,
).pop().title;
