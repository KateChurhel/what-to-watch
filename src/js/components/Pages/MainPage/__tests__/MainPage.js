import React from 'react';
import { render } from '@testing-library/react';
import MainPage from '../index';
import { MOCK_MOVIES, MOCK_TVS } from '../../../../mocks/movies';
import { formatCardData } from '../../../../helpers/formatData';

describe('Main Page: ', () => {
  it('correctly renders', async () => {
    const { asFragment } = render(
      <MainPage
        bigMovieCardData={formatCardData(MOCK_MOVIES[0])}
        mostPopularMovies={MOCK_MOVIES.slice(0, 2)}
        mostPopularTvs={MOCK_TVS.slice(0, 2)}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders without data', async () => {
    const { asFragment } = render(<MainPage />);

    expect(asFragment()).toMatchSnapshot();
  });
});
