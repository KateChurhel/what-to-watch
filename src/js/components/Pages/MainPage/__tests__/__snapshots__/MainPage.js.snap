// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page:  correctly renders 1`] = `
<DocumentFragment>
  <section
    class="movie-card"
  >
    <div
      class="movie-card__bg"
      style="background-image: url(https://image.tmdb.org/t/p/original/qJeU7KM4nT2C1WpOrwPcSDGFUWE.jpg);"
    />
    <div
      class="container"
    >
      <div
        class=" movie-card__wrap"
      >
        <div
          class="movie-card__info movie-card__info--main"
        >
          <div
            class="movie-card__poster movie-card__poster--main"
          >
            <img
              alt="La La Land"
              loading="lazy"
              src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
            />
          </div>
          <div
            class="movie-card__desc"
          >
            <h2
              class="movie-card__title"
            >
              La La Land (2016)
            </h2>
            <p
              class="movie-card__meta"
            >
              <span
                class="movie-card__genre"
              >
                Comedy, Drama, Romance, Music
              </span>
              <span
                class="movie-card__runtime"
              >
                2h 9m
              </span>
            </p>
            <i
              class="movie-card__tagline movie-card__tagline--main"
            >
              Here's to the fools who dream.
            </i>
            <a
              class="button movie-card__button"
              href="/details/movie/313369"
            >
              More Info
            </a>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="container"
  >
    <div
      class="swiper__title-wrapper"
    >
      <h2
        class="swiper__title"
      >
        Most Popular Movies
      </h2>
      <a
        href="/movie/1"
      >
        Show All
      </a>
    </div>
    <div
      class="swiper-container swiper-container-initialized swiper-container-horizontal"
    >
      <div
        class="swiper-button-prev"
      />
      <div
        class="swiper-button-next"
      />
      <div
        class="swiper-wrapper"
        style="transition: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
      >
        <div
          class="swiper-slide"
        >
          <a
            class="small-movie-card catalog__movies-card small-movie-card__link"
            href="/details/movie/313369"
          >
            <div
              class="small-movie-card__image "
            >
              <img
                alt="La La Land"
                height="450"
                loading="lazy"
                src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
                width="300"
              />
            </div>
            <div
              class="small-movie-card__info"
            >
              <h3
                class="small-movie-card__title"
              >
                La La Land
              </h3>
            </div>
          </a>
        </div>
        <div
          class="swiper-slide"
        >
          <a
            class="small-movie-card catalog__movies-card small-movie-card__link"
            href="/details/movie/516486"
          >
            <div
              class="small-movie-card__image "
            >
              <img
                alt="Greyhound"
                height="450"
                loading="lazy"
                src="https://image.tmdb.org/t/p/w300/kjMbDciooTbJPofVXgAoFjfX8Of.jpg"
                width="300"
              />
            </div>
            <div
              class="small-movie-card__info"
            >
              <h3
                class="small-movie-card__title"
              >
                Greyhound
              </h3>
            </div>
          </a>
        </div>
      </div>
    </div>
    <div
      class="swiper__title-wrapper"
    >
      <h2
        class="swiper__title"
      >
        Most Popular TV Shows
      </h2>
      <a
        href="/tv/1"
      >
        Show All
      </a>
    </div>
    <div
      class="swiper-container swiper-container-initialized swiper-container-horizontal"
    >
      <div
        class="swiper-button-prev"
      />
      <div
        class="swiper-button-next"
      />
      <div
        class="swiper-wrapper"
        style="transition: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
      >
        <div
          class="swiper-slide"
        >
          <a
            class="small-movie-card catalog__movies-card small-movie-card__link"
            href="/details/tv/1622"
          >
            <div
              class="small-movie-card__image "
            >
              <img
                alt="Supernatural"
                height="450"
                loading="lazy"
                src="https://image.tmdb.org/t/p/w300/KoYWXbnYuS3b0GyQPkbuexlVK9.jpg"
                width="300"
              />
            </div>
            <div
              class="small-movie-card__info"
            >
              <h3
                class="small-movie-card__title"
              >
                Supernatural
              </h3>
            </div>
          </a>
        </div>
        <div
          class="swiper-slide"
        >
          <a
            class="small-movie-card catalog__movies-card small-movie-card__link"
            href="/details/tv/95"
          >
            <div
              class="small-movie-card__image "
            >
              <img
                alt="Buffy the Vampire Slayer"
                height="450"
                loading="lazy"
                src="https://image.tmdb.org/t/p/w300/41jsOEYsh9lJBc6AZqGpR1ETpx5.jpg"
                width="300"
              />
            </div>
            <div
              class="small-movie-card__info"
            >
              <h3
                class="small-movie-card__title"
              >
                Buffy the Vampire Slayer
              </h3>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Main Page:  correctly renders without data 1`] = `
<DocumentFragment>
  <div
    class="container"
  />
</DocumentFragment>
`;
